code/notificationbuilder/src/test/java/com/adobe/marketing/mobile/notificationbuilder/internal/builders/ProductCatalogNotificationBuilderTest.kt

```diff
@@ -0,0 +1,147 @@
+import com.adobe.marketing.mobile.notificationbuilder.internal.templates.MockProductCatalogTemplateDataProvider.getMockedMapWithProductCatalogData
+import com.adobe.marketing.mobile.notificationbuilder.internal.templates.ProductCatalogPushTemplate
+import com.adobe.marketing.mobile.notificationbuilder.internal.templates.provideMockedProductCatalogTemplate
+import com.adobe.marketing.mobile.notificationbuilder.internal.templates.replaceValueInMap
+import com.adobe.marketing.mobile.notificationbuilder.internal.util.MapData
+import io.mockk.every
+import io.mockk.mockkClass
+import io.mockk.mockkObject
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.mockito.Mock
+import org.mockito.MockitoAnnotations
+import org.robolectric.RobolectricTestRunner
+import org.robolectric.RuntimeEnvironment
+import org.robolectric.annotation.Config
+import kotlin.test.assertEquals
+import kotlin.test.assertFailsWith
+
+@RunWith(RobolectricTestRunner::class)
+@Config(sdk = [31])
+class ProductCatalogNotificationBuilderTest {
+    @Mock
+    private lateinit var context: Context
+    private lateinit var trackerActivityClass: Class<out Activity>
+    private lateinit var broadcastReceiverClass: Class<out BroadcastReceiver>
+
+    @Before
+    fun setUp() {
+        MockitoAnnotations.openMocks(this)
+
+        context = RuntimeEnvironment.getApplication()
+        trackerActivityClass = DummyActivity::class.java
+        broadcastReceiverClass = DummyBroadcastReceiver::class.java
+        mockkObject(PushTemplateImageUtils)
+    }
+
+    @Test
+    fun `construct should throw NotificationConstructionFailedException if downloaded image count does not match catalog image count`() {
+        val pushTemplate = provideMockedProductCatalogTemplate()
+
+        assertFailsWith(
+            exceptionClass = NotificationConstructionFailedException::class,
+            message = "Failed to download all images for the product catalog notification.",
+            block = {
+                ProductCatalogNotificationBuilder.construct(context, pushTemplate, trackerActivityClass, broadcastReceiverClass)
+            }
+        )
+    }
+
+    @Test
+    fun `construct should return a NotificationCompat Builder if download image count matches catalog image count`() {
+        val pushTemplate = provideMockedProductCatalogTemplate()
+        val cachedItem = mockkClass(Bitmap::class)
+
+        // product catalog template requires 3 catalog items
+        every { cacheImages(any()) } answers { 3 }
+        every { getCachedImage(any()) } answers { cachedItem }
+        val notificationBuilder = ProductCatalogNotificationBuilder.construct(context, pushTemplate, trackerActivityClass, broadcastReceiverClass)
+
+        assertEquals(R.layout.push_template_horizontal_catalog, notificationBuilder.bigContentView.layoutId)
+        assertEquals(NotificationCompat.Builder::class.java, notificationBuilder::class.java)
+    }
+
```

Comment:
can we add a test which verifies an exception is thrown if we have more than 3 catalog items?